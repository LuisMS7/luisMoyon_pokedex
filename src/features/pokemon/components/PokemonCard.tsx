import React, { useEffect, useMemo, useState } from "react";
import { Pokemon } from "@pokemon/types/pokemon.type";
import { StyleSheet, Text, View } from "react-native";
import SvgUri from "react-native-svg-uri";
import { TypeColor } from "@pokemon/utils/color-by-type";
import { perfectSize } from "@utils/perfect-size";
import { TEXT_SUBTITLE_SIZE, TEXT_TITLE_SIZE } from "@lib/constants";
import { titlecase } from "@utils/string-formatter";
// eslint-disable-next-line max-len
import { PokemonListItemApiResponse } from "@pokemon/types/pokemon-list-api-response.type";
import { useApi } from "@hooks/useApi";
import getPokemonByName from "@pokemon/store/actions/get-pokemon-by-name";
import { PokemonApiResponse } from "@pokemon/types/pokemon-api-response.type";

type PokemonCardType = {
	pokemonItem: PokemonListItemApiResponse;
};
const PokemonCard = ({ pokemonItem }: PokemonCardType) => {
	const [pokemon, setPokemon] = useState<Pokemon>();
	const { request, response } = useApi<PokemonApiResponse>(() =>
		getPokemonByName(pokemonItem.name),
	);

	useEffect(() => {
		request();
	}, []);

	useEffect(() => {
		if (response && !pokemon) {
			const pokemonFetched: Pokemon = {
				id: response.id,
				name: response.name,
				img:
					response?.sprites?.other?.dream_world?.front_default ||
					response?.sprites?.front_default,
				types: response.types,
				weight: response.weight,
				sprites: response.sprites,
				moves: response.moves,
			};
			setPokemon(pokemonFetched);
		}
	}, [response]);

	const typeColor = useMemo(() => {
		return pokemon ? TypeColor[pokemon.types[0].type.name] : "white";
	}, [pokemon]);

	return (
		<View style={{ ...styles.container, borderColor: typeColor }}>
			<Text style={{ ...styles.idText, color: typeColor }}>
				#{pokemon?.id.toString().padStart(4, "0")}
			</Text>
			<View style={styles.image}>
				<SvgUri
					source={{ uri: pokemon?.img }}
					height={perfectSize(250)}
				/>
			</View>
			<Text style={{ ...styles.nameText, backgroundColor: typeColor }}>
				{titlecase(pokemon?.name.toUpperCase())}
			</Text>
		</View>
	);
};

const styles = StyleSheet.create({
	container: {
		width: perfectSize(350),
		height: "100%",
		borderRadius: perfectSize(16),
		borderWidth: perfectSize(4),
		flexDirection: "column",
	},
	idText: {
		width: "100%",
		textAlign: "right",
		padding: perfectSize(20),
		fontSize: TEXT_SUBTITLE_SIZE,
	},
	image: {
		flex: 1,
		alignItems: "center",
	},
	nameText: {
		fontSize: TEXT_TITLE_SIZE,
		color: "white",
		paddingVertical: perfectSize(20),
		textAlign: "center",
	},
});
export default PokemonCard;
