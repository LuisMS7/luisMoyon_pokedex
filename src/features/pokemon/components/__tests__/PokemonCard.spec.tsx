import "react-native";
import React from "react";
import { render, screen } from "@testing-library/react-native";
import PokemonCard from "@pokemon/components/PokemonCard";
import { Pokemon } from "@pokemon/types/Pokemon.type";

describe("PokemonCard", () => {
	it("should render correctly", () => {
		const mockPokemon: Pokemon = {
			name: "bulbasaur",
			id: 1,
			// eslint-disable-next-line max-len
			img: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg",
			types: [
				{
					slot: 1,
					type: {
						name: "grass",
						url: "https://pokeapi.co/api/v2/type/12/",
					},
				},
				{
					slot: 2,
					type: {
						name: "poison",
						url: "https://pokeapi.co/api/v2/type/4/",
					},
				},
			],
		};
		render(<PokemonCard pokemon={mockPokemon} />);
		expect(screen.toJSON()).toMatchSnapshot();
	});
});
