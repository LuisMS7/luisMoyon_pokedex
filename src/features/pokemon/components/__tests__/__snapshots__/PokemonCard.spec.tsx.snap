// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonCard should render correctly 1`] = `
<View
  style={
    {
      "borderColor": "#74CB48",
      "borderRadius": 23.855932837144906,
      "borderWidth": 5.963983209286226,
      "flexDirection": "column",
      "height": "100%",
      "width": 521.8485308125448,
    }
  }
>
  <Text
    style={
      {
        "color": "#74CB48",
        "fontSize": 29.819916046431132,
        "padding": 29.819916046431132,
        "textAlign": "right",
        "width": "100%",
      }
    }
  >
    #
    0001
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
      }
    }
  />
  <Text
    style={
      {
        "backgroundColor": "#74CB48",
        "color": "white",
        "fontSize": 35.78389925571736,
        "paddingVertical": 29.819916046431132,
        "textAlign": "center",
      }
    }
  >
    Mockpokemon
  </Text>
</View>
`;
